bluefoglite/common/optimizers.py

```diff
@@ -0,0 +1,234 @@
+                                == CommunicationType.neighbor_allreduce
+                            ):
+                                async_work = self._neighbor_allreduce_data_async(p)
+                            elif self._communication_type == CommunicationType.empty:
+                                async_work = None
+                            else:
+                                raise ValueError(
+                                    "Unsuppported CommunicationType encountered."
+                                )
+                            self._async_works[p] = async_work
+
+        return hook
+
+    def _neighbor_allreduce_data_async(self, p):
+        async_work = bfl.neighbor_allreduce_nonblocking(
+            p.data,
+            self_weight=self.self_weight,
+            src_weights=self.src_weights,
+            dst_weights=self.dst_weights,
+        )
+        return async_work
+
+    def _allreduce_data_async(self, p):
+        async_work = bfl.allreduce_nonblocking(p.data)
+        return async_work
+
+    @property
+    def communication_type(self):
+        return self._communication_type
+
+    @communication_type.setter
+    def communication_type(self, value):
+        assert isinstance(value, CommunicationType)
+        self._communication_type = value
+
+    def synchronize(self):
+        with torch.no_grad():
+            for p, async_work in self._async_works.items():
+                if async_work is not None:
+                    output = async_work.wait()
+                    p.set_(output)
+                # self._async_works[p] = self._num_steps_per_communication
+                self._reduce_delay[p] = self._num_steps_per_communication
+        self._async_works.clear()
+        self._synchronized = True
+
+    def step(self, closure=None):
+        # consensus style is the easiest way to implement it.
+        if self._should_synchronize:
+            if self._synchronized:
+                warnings.warn(
+                    "optimizer.step() called without "
+                    "optimizer.skip_synchronize() context after "
+                    "optimizer.synchronize(). This can cause training "
+                    "slowdown. You may want to consider using "
+                    "optimizer.skip_synchronize() context if you use "
+                    "optimizer.synchronize() in your code."
+                )
+            self.synchronize()
+        self._synchronized = False
+        return super(self.__class__, self).step(closure)
+
+
+def DistributedAdaptWithCombineOptimizer(
+    optimizer,
+    model,
+    communication_type=CommunicationType.neighbor_allreduce,
+    num_steps_per_communication=1,
+):
+    cls = type(
+        optimizer.__class__.__name__,
+        (optimizer.__class__,),
+        dict(_DistributedReduceOptimizer.__dict__),
+    )
+    return cls(
+        optimizer.param_groups, model, communication_type, num_steps_per_communication
+    )
```

Comment:
I think now this trick no longer needed... just normal inherit should be sufficient